### PR TITLE
Add a class to manage desktop screenshots 

### LINKED ISSUE #3: Daemon that will screenshot desktops and expose them through socket interface
The daemon has to
- Make and store Pixmaps of each virtual desktop together with the _NET_DESKTOP_NAME
- Send Pixmaps to GUI client over unix sockets
  - Make a socket server (daemon) and a socket client (GUI)

### PR DESCRIPTION
**Pull request by mmskv**
Now it only
* Saves preset screen region to specified image format
* Creates new pixmap and copies saved image onto it

Need code review for further development

### CODE REVIEW

File: src/DesktopPixmap.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    const uint16_t height,  ///< height of display
+    const uint16_t width,   ///< width of display
+    const std::string &name ///< _NET_WM_NAME of display
+)
+{
+  this->x = x;
+  this->y = y;
+  this->height = height;
+  this->width = width;
+  this->name = name; // XXX Is this OK?
```
Comment (TheHussky) [inline src/DesktopPixmap.cpp]: I'm not sure that's ok. Name is a reference itself(why, btw?), a pointer to reference might lead to some unexpected errors. Moreover, the <type & variable> is used in args when it needs to be changed not only within the function, but globally.

File: src/DesktopPixmap.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  DesktopPixmap (short x,         ///< x coordinate of the top left corner
+                 short y,         ///< y coordinate of the top left corner
+                 uint16_t height, ///< Height of the display
+                 uint16_t width,  ///< Width of the display
+                 const std::string &name ///< Name of the display (_NET_WM_NAME)
+  );
+  ~DesktopPixmap ();
+
+  /**
+   * TODO Rename to something more fitting
```
Comment (TheHussky) [inline src/DesktopPixmap.hpp]: I suggest naming it smth like "save_screen_copy"

File: src/DesktopPixmap.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+   */
+  void saveImage (xcb_connection_t *c, xcb_screen_t &screen,
+                  PixmapFormat format = kZPixmap // Default to more efficient
+  );
+  void putImage (xcb_connection_t *c, xcb_screen_t &screen);
+
+private:
+  // XXX Do I need static here?
```
Comment (TheHussky) [inline src/DesktopPixmap.hpp]: static methods are used to use class without initializing the object of this class. You can create gc without using this class at all, so not really, I guess. Not quite sure though, tbh.

File: src/DesktopPixmap.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    {
+      create_pixmap (c, screen);
+    }
+
+  this->format = pixmap_format;
+
+  auto gi_cookie = xcb_get_image (
+      c,            /* Connection */
+      this->format, /* Image format */
+      screen.root,  /* Screenshot relative to root */
```
Comment (TheHussky) [inline src/DesktopPixmap.cpp:55]: isn't it screen->root?

File: src/DesktopPixmap.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  DesktopPixmap::gc_ = 0;
+  this->image_ptr = nullptr;
+  this->pixmap_id = 0;
+  this->length = 0;
+}
+
+DesktopPixmap::~DesktopPixmap ()
+{
+  delete[] this->image_ptr;
+  // TODO Might need to delete pixmap
```
Comment (TheHussky) [inline src/DesktopPixmap.cpp]: You can set it free via xcb_free_pixmap anytime, I think. If it's not what you mean by delete, then what?

File: src/DesktopPixmap.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                 screen.root_depth, this->length, this->image_ptr);
+}
+
+void
+DesktopPixmap::create_gc (xcb_connection_t *c, xcb_screen_t &screen)
+{
+  uint32_t mask;
+  uint32_t values[2];
+
+  // TODO Figure out correct mask and values
```
Comment (TheHussky) [inline src/DesktopPixmap.cpp]: I have seen source code for masks somewhere, their names speak for themselves sometimes. I suggest using transparent background by the way.

File: src/DesktopPixmap.cpp (hunk 5 above)
Comment (mmskv) [reply src/DesktopPixmap.cpp]: Yes, you are right. Will have to free pixmap in the destructor

File: src/DesktopPixmap.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  DesktopPixmap::gc_ = xcb_generate_id (c);
+  xcb_create_gc (c, DesktopPixmap::gc_, screen.root, mask, values);
+}
+
+void
+DesktopPixmap::create_pixmap (xcb_connection_t *c, xcb_screen_t &screen)
+{
+  this->pixmap_id = xcb_generate_id (c);
+  xcb_create_pixmap (c, screen.root_depth, this->pixmap_id,
+                     screen.root, // TODO Check if this is OK
```
Comment (TheHussky) [inline src/DesktopPixmap.cpp]: Are you afraid of segmentation faults? Seemed alright when I tested smth. The screen->root use never caused errors, if I remember correctly. You use &screen, might fail, but I don't think chaces are high.

**Review by mmskv — COMMENTED** [review]
Have to link this class in CMakeLists.txt

File: src/DesktopPixmap.cpp (hunk 7 above)
Comment (mmskv) [reply src/DesktopPixmap.cpp]: In documentation for pixmaps it says, that `drawable` argument specifies screen, outside of which the pixmap can't be used. LGTM, as pixmaps won't be used on a different screen.

File: src/DesktopPixmap.cpp (hunk 6 above)
Comment (mmskv) [reply src/DesktopPixmap.cpp]: Can you link me the docs for masks and values please

File: src/DesktopPixmap.cpp (hunk 1 above)
Comment (mmskv) [reply src/DesktopPixmap.cpp]: Looks like it does not matter for short strings.

As I won't change the passed `const` string, only move operations will be called by the compiler. But code generated with `-O3` flag looks the same both for the const reverence and by-value argument passes.

File: src/DesktopPixmap.hpp (hunk 2 above)
Comment (mmskv) [reply src/DesktopPixmap.hpp]: Too long imo
And `screen` doesn't fit, as this is a different concept in X11

**Comment by mmskv** [comment]
Should I create a `xcb_connection_t` for each DesktopPixmap instance?
Passing `*c` everywhere is tedious. And I can't free pixmap without `xcb_connection_t` in the destructor.
But maybe it takes up too many resources, need to look into it.

I think the best move here is to create an `inline static xcb_connection_t c` for the class.

**Review by TheHussky — APPROVED** [review]
LGTM. I'll think about a proper name to save_image.

File: src/DesktopPixmap.hpp (hunk 3 above)
Comment (mmskv) [reply src/DesktopPixmap.hpp]: https://stackoverflow.com/a/6446025
Yes, declaring this function as a class member is redundant, but it looks good, so why not

File: src/DesktopPixmap.cpp (hunk 5 above)
Comment (mmskv) [reply src/DesktopPixmap.cpp]: If I do `xcb_disconnect`, will every x object allocated on server be freed?
Sounds reasonable

File: src/DesktopPixmap.cpp (hunk 6 above)
Comment (mmskv) [reply src/DesktopPixmap.cpp]: https://www.x.org/releases/current/doc/libxcb/tutorial/index.html

**Review by mmskv — COMMENTED** [review]
Nice job!
Couple of notes:
* I see you don't have an autoformatter configured. It is essential, because then my autoformatter won't edit your lines of code and mess up commits. So these things need to be changed:
  * Add a whitespace before and after comments (my c++ formatter does this automatically, perhaps you should configure yours too)
  * Separate function definitions with newline (add newline between l126 and l127)
  * Add a newline at the end of the file
  * Use consistent padding around operators. See .clang-format
  * Don't exceed 80 lines of code width. See .clang-tidy
* `char flag` looks ugly imo, i'm sure there is a better way to implement this 

I'll setup a code style pipeline in the meantime

**Review by kud-aa — APPROVED** [review]

**Pull request merged**